src/core/channelz/channelz_registry.cc

```diff
@@ -247,95 +242,75 @@ ChannelzRegistry::ShardedNodeMap::QueryNodes(
-      if (node_ref == nullptr) {
-        n = n->next_;
-        continue;
-      }
-      BaseNode* next = n->next_;
-      RemoveNodeFromHead(n, node_shard.nursery);
-      AddNodeToHead(n, node_shard.numbered);
-      n->uuid_ = uuid_generator_;
-      ++uuid_generator_;
-      index_.emplace(n->uuid_, n);
-      if (n->uuid_ >= start_node) {
-        if (result.size() == max_results) {
-          node_after_end = std::move(node_ref);
-          return std::tuple(std::move(result), false);
+    for (auto [nursery, numbered] :
+         {std::pair(&node_shard.nursery, &node_shard.numbered),
+          std::pair(&node_shard.orphaned, &node_shard.orphaned_numbered)}) {
+      if (nursery->head == nullptr) continue;
+      BaseNode* n = nursery->head;
+      while (n != nullptr) {
+        if (!discriminator(n)) {
+          n = n->next_;
+          continue;
         }
-        result.emplace_back(std::move(node_ref));
+        auto node_ref = n->WeakRefIfNonZero();
+        if (node_ref == nullptr) {
+          n = n->next_;
+          continue;
+        }
+        BaseNode* next = n->next_;
+        nursery->Remove(n);
+        numbered->AddToHead(n);
+        n->uuid_ = uuid_generator_;
+        ++uuid_generator_;
+        index_.emplace(n->uuid_, n);
+        if (n->uuid_ >= start_node) {
+          if (result.size() == max_results) {
+            node_after_end = std::move(node_ref);
+            return std::tuple(std::move(result), false);
+          }
+          result.emplace_back(std::move(node_ref));
+        }
+        n = next;
       }
-      n = next;
     }
   }
   CHECK(node_after_end == nullptr);
   return std::tuple(std::move(result), true);
 }
 
-void ChannelzRegistry::ShardedNodeMap::AddNodeToHead(BaseNode* node,
-                                                     BaseNode*& head) {
-  if (head == nullptr) {
-    head = node;
-    node->prev_ = node->next_ = nullptr;
-    return;
-  }
-  DCHECK_EQ(head->prev_, nullptr);
-  node->next_ = head;
-  node->prev_ = nullptr;
-  head->prev_ = node;
-  head = node;
-}
-
-void ChannelzRegistry::ShardedNodeMap::RemoveNodeFromHead(BaseNode* node,
-                                                          BaseNode*& head) {
-  if (node == head) {
-    DCHECK_EQ(node->prev_, nullptr);
-    head = node->next_;
-    if (head != nullptr) head->prev_ = nullptr;
-    return;
-  }
-  node->prev_->next_ = node->next_;
-  if (node->next_ != nullptr) node->next_->prev_ = node->prev_;
-}
-
-WeakRefCountedPtr<BaseNode> ChannelzRegistry::ShardedNodeMap::GetNode(
-    intptr_t uuid) {
+WeakRefCountedPtr<BaseNode> ChannelzRegistry::InternalGet(intptr_t uuid) {
   MutexLock index_lock(&index_mu_);
   auto it = index_.find(uuid);
   if (it == index_.end()) return nullptr;
   BaseNode* node = it->second;
   return node->WeakRefIfNonZero();
 }
 
-intptr_t ChannelzRegistry::ShardedNodeMap::NumberNode(BaseNode* node) {
+intptr_t ChannelzRegistry::InternalNumberNode(BaseNode* node) {
```

Comment:
I don't see this being called from anywhere.  What is the use-case for this?

Reply:
It's called from `ChannelRegistry::NumberNode`, which is called from `BaseNode::UuidSlow` (which is called from `BaseNode::uuid`) -- the use case is that some code needs to refer to the uuid of a node outside of the query system - say for returning the list of child subchannels. In those cases we need a path to force generation of a uuid.